src/Interpreters/ExecuteScalarSubqueriesVisitor.cpp

```diff
@@ -72,40 +79,95 @@ static bool worthConvertingToLiteral(const Block & scalar)
     return !useless_literal_types.count(scalar_type_name);
 }
 
+static auto getQueryInterpreter(const ASTSubquery & subquery, ExecuteScalarSubqueriesMatcher::Data & data)
+{
+    auto subquery_context = Context::createCopy(data.getContext());
+    Settings subquery_settings = data.getContext()->getSettings();
+    subquery_settings.max_result_rows = 1;
+    subquery_settings.extremes = false;
+    subquery_context->setSettings(subquery_settings);
+    if (subquery_context->hasQueryContext())
+    {
+        /// Save current cached scalars in the context before analyzing the query
```

Comment:
This helps boost the HIT ratio in queries with several CTEs and the default `enable_global_with_statement` as they always chain the previous subqueries to themselves. For example:

```
WITH
    ( SELECT sleep(0.0001) FROM system.one ) as a1,
    ( SELECT sleep(0.0001) FROM system.one ) as a2,
    ( SELECT sleep(0.0001) FROM system.one ) as a3,
    ( SELECT sleep(0.0001) FROM system.one ) as a4,
    ( SELECT sleep(0.0001) FROM system.one ) as a5
```

First a1 is calculated. Then a2 (with a1 attached) is calculated. Then a3 (with a2, a1 attached) is calculated.

By pushing a1 value to the context before calculating a2 we avoid calculating it again. When pushing a3 we now avoid calculating a2 (which before would mean calculating a1 too) and so on.